source/extensions/filters/common/rbac/engine.h

```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "envoy/http/filter.h"
+#include "envoy/http/header_map.h"
+#include "envoy/network/connection.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Filters {
+namespace Common {
+namespace RBAC {
+
+/**
+ * Shared logic for evaluating RBAC policies for both network and HTTP filters.
+ */
+class RBACEngine : public Router::RouteSpecificFilterConfig {
+public:
+  virtual ~RBACEngine() {}
+
+  /**
+   * Returns whether or not the current action is permitted. This overload of `allowed` should only
+   * be used for non-HTTP network calls (e.g., a binary TCP protocol).
+   *
+   * @param connection the downstream connection used to identify the action/principal.
+   */
+  virtual bool allowed(const Network::Connection& connection) const PURE;
```

Comment:
This comment is not going to make you happy, but I'm wondering what you think about maintaining the bifurcated variants of allowed/matches through this code. If you look at what we did in ratelimit, we basically use the same call points whether it's HTTP or not (similar for access logging). There are two ways to do this:
1) Allow headers to be nullptr (switch to const raw pointer)
2) Synthetically pass a static const empty http headers for calls from the eventual network filter. This makes the matching code even simpler since there needs to be no null checking anywhere in the engine.

I would vote that we get rid of the split functions everywhere, and do number 2 above. Given this PR just touches HTTP, it will always be populated with a real header block. The eventual network filter can just pass a static empty head block as I already mentioned. I think this will delete a bunch of code. Thoughts?

Reply:
Going with (2). Easy enough!